blackjax/tempered_smc.py

```diff
@@ -0,0 +1,116 @@
+    lamda: float
+
+
+class SMCInfo(NamedTuple):
+    """Additional information on the tempered SMC step.
+
+    weights: jnp.ndarray
+        The weights after the MCMC pass.
+    """
+    weights: jnp.ndarray
+
+
+new_state = base.new_hmc_state
+
+
+def _normalize(log_weights):
+    """ Normalize log-weights into weights
+    """
+    w = jnp.exp(log_weights - jnp.max(log_weights))
+    return w / w.sum()
+
+
+def kernel(potential_fn: Callable, mcmc_kernel: Callable, new_mcmc_state: Callable, resampling_method: Callable,
+           mcmc_parameters: HMCParameters, mcmc_steps=10):
+    """Build a Tempered SMC step.
+
+    Parameters
+    ----------
+    potential_fn: Callable
+        A function that returns the potential energy of a chain at a given position.
+    mcmc_kernel: Callable
+        A callable function that creates a kernel from `potential_function` and parameters `mcmc_parameters`
+    new_mcmc_state: Callable
+        How to create a new MCMC state for the MCMC routine to start
+    resampling_method: Callable
+        A random function that resamples generated particles based of weights
+    mcmc_parameters: HMCParameters
+        The parameters to be fed to the mcmc kernel
+    mcmc_steps: int
+        Number of MCMC steps to run per SMC step
+
+    Returns
+    -------
+    A callable that takes a rng_key and a Pytree that contains the current state
+    of the chain and that returns a new state of the chain along with
+    information about the transition.
+    """
+
+    def one_step(rng_key: jnp.ndarray, state: SMCState, delta: float) -> Tuple[SMCState, SMCInfo]:
+        """
+
+        Parameters
+        ----------
+        rng_key: DeviceArray[int],
+            JAX PRNGKey for randomness
+        state: SMCState
+            Current state of the tempered SMC algorithm
+        delta: float
+            Increment for `state.lamda
+
+        Returns
+        -------
+        state: SMCState,
+            The new state of the tempered SMC algorithm
+         info: SMCInfo,
+            Additional information on the SMC step
+        """
+        n_particles = jax.tree_flatten(state.particles)[0].shape[0]
+
+        def tempered_potential_fn(pytree):
+            return -state.lamda * potential_fn(pytree)
+
+        step_mcmc_kernel = jax.vmap(mcmc_kernel(tempered_potential_fn, mcmc_parameters), in_axes=(0, 0))
+        rng_key, resampling_key = jax.random.split(rng_key, 2)
+        keys = jax.random.split(rng_key, mcmc_steps * n_particles)
+        keys = keys.reshape((mcmc_steps, n_particles))
+
+        def mcmc_loop_body(mcmc_state, body_keys):
+            mcmc_state, _ = step_mcmc_kernel(body_keys, mcmc_state)
+            return mcmc_state, None
+
+        initial_mcmc_states = jax.vmap(new_mcmc_state, in_axes=(0, None))(state.particles, tempered_potential_fn)
+        last_mcmc_state, _ = jax.lax.scan(mcmc_loop_body, initial_mcmc_states, keys)
+        particles = last_mcmc_state.position
```

Comment:
Let me try to simplify this a little bit:

```python
mcmc_key, resampling_key = jax.random.split(rng_key, 2)

def mcmc_step(states, rng_key):
    keys = jax.random.split(rng_key, num_particules)
    states, info = jax.vmap(kernel)(states)
    return states, (states, info)

keys = jax.random.split(mcmc_key, num_mcmc_steps)
initial_states = jax.vmap(new_mcmc_state, in_axes=(0, None))(state.particles, tempered_potential_fn)
last_states, _ = jax.lax.scan(mcmc_step, initial_states, keys)
```

Reply:
I am not a fan of splitting keys within a loop, it is less efficient for very little added visibility. Also in the case of tempered SMC, there is no reason to store the intermediate states of the MCMC chain like you just did, (except for waste free I guess, but the implementation would be wildly different)

Reply:
More interestingly what do you think of the vmap here. Should it be the responsibility of the user (this way we allow for manual batching), or abstracted away in the code like I've done? It's pretty hard to catch the fact that the kernel may not be batched, so if we leave it as a user responsibility, returning readable errors saying "hey you forgot to batch" is probably gonna be impossible

Reply:
- I haven't noticed much difference between splitting all at once and splitting within the loop, but I haven't tried in a while.
- We'll come to users batching themselves later I think; you wouldn't batch NUTS like you do HMC for instance, since NUTS chains can take very different number of steps at each round. For now we can batch inside as I don't know what the API will look like.